Add post about performance benefits of opaque types (#23)
elmcraft/elmcraft.org: content/learn/opaque-types-performance.md

---
type: page
title: A case for opaque types

> description: The performance benefits of defaulting to opaque types

Reviewer: I think `Compilation performance benefits of opaque types` might be a better title here. The description could then be "Exploring the case for defaulting to opaque types"

Reviewer 2: (I have yet to read the article but) these are not opaque types per se, they're custom types (maybe "wrapper" custom types, not sure). Opaque types can by definition not be exposed, as that would allow the importer module to the internals of the custom type. But for the sake of this performance issue, it's fine to expose the types.

published: false
status: seedling
authors: "@aemengo"
editors: ""
---

> [Opaque types](https://sporto.github.io/elm-patterns/advanced/opaque-types.html) as an Elm design pattern always made sense - for libraries. The idea being that I use a **union type** for a plain old Elm object, instead of an **alias type**, even when there's only one **variant**.

Reviewer 2: My feeling (but feel free to disregard it) is that the post uses maybe an unoptimal route for explaining the problem and the solution, going through explaining opaque types which is not the main concern. You (and the people who will be sent this link) have a **compilation** performance issue which is resolved through custom types.

I think it would be more to the point to change the article to talk about potential issues around compilation speeds, things to look for, and solutions. I would really emphasize **compilation** speeds too, and maybe not talk about design and custom types which I would say are a bit beside the point.

I think the style chosen for this article is also more of a blog post (see the uses of the first pronoun and the usage of terms like "assume") rather than documentation where people are looking for solutions to their problems (See for example this article for the style I'm thinking of: https://elmcraft.org/compare/javascript/eslint).

So from:

```
type alias Model = { … }
```

to

```
type Model = Model { … }
```

> I would do this be able to **expose** the `type Model`, but not the `Model {...}`.<br/>

Reviewer 2: "I would do this be able to" -> "I would do this TO be able to"?

That way I (*the library author*) can change `Model {...}` without you (*the one using my library*) having to change all your code.

"Well... my Elm project isn't a library, so I don't need this 🤷", I thought to myself.

## But I was wrong...

The issues started when my CI system consistently failed to build my Elm project.

![CI kills Elm build](./img/otp-ci-kill-elm-build.png)

> A little sleuthing lead me to find that my CI system was actually killing the `elm make` process because it was consuming too much memory. I was incredulous at first but here's the make process again in a local docker image, with a 4GB memory limit.

Reviewer 2: "lead" -> "led"

![Elm build in docker memory graph](./img/otp-docker-memory-graph.png)

The make process spikes the memory all the way to the cap at 4GB, and is subsequently killed.

Running the following command, lead me to believe that it was my liberal usage of `type alias` contributing to excessive amounts of garbage collection during the `elm make` process.

```
# inside source code folder
du -hs elm-stuff/0.19.1/* | sort -h | tail -n 30 | tac
```

![Elm stuff content](./img/otp-elm-stuff-content.png)

> The helpful folk at the [Elm Slack Community](https://elm-lang.org/community) taught me that `.elmi` files `>10MB` are considered _extremely_ large. Even `>1MB` was pushing it.

Reviewer 2: ```suggestion
The helpful folks on the [Elm Slack](https://elm-lang.org/community) taught me that `.elmi` files `>10MB` are considered _extremely_ large. Even `>1MB` was pushing it.
```

## The neat thing...

As soon as I converted my 3 most prominent **alias types** to the **opaque types** mentioned earlier, all the issues went away: the build failures, the memory spikes, and the slow compilation speeds.

> I'd assume that the compiler has to perform double duty (*checking all the fields*) every time that an **alias type** must be passed into a method invocation. This being because any object would qualify as the **alias type** so long as all the fields match. There doesn't seem to be any official best practice about this in their docs...

Reviewer 2: Would be nice if we could figure out why exactly, rather than assume. I don't know the answer though.
